railgun/controllers/configuration/secret_utils.go

```diff
@@ -0,0 +1,166 @@
+package configuration
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	"github.com/kong/kubernetes-ingress-controller/railgun/controllers"
+	"github.com/kong/kubernetes-ingress-controller/railgun/pkg/configsecret"
+)
+
+// -----------------------------------------------------------------------------
+// Secret Utils - Storage
+// -----------------------------------------------------------------------------
+
+// storeIngressObj reconciles storing the YAML contents of Ingress resources (which are managed by Kong)
+// from multiple versions which remain supported.
+func storeIngressObj(ctx context.Context, c client.Client, log logr.Logger, nsn types.NamespacedName, obj client.Object) (ctrl.Result, error) {
+	// TODO need EVENTS here
+	// TODO need more status updates
+	// TODO: (shane) I want to refactor this into several smaller functions
+	// ^ follow up for these items is in: https://github.com/Kong/kubernetes-ingress-controller/issues/1094
+
+	// if this is an Ingress resource make sure it's managed by Kong
+	if obj.GetObjectKind().GroupVersionKind().Kind == "Ingress" {
+		if !isManaged(obj.GetAnnotations()) {
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// get the configuration secret namespace
+	secretNamespace := os.Getenv(controllers.CtrlNamespaceEnv)
+	if secretNamespace == "" {
+		return ctrl.Result{}, fmt.Errorf("kong can not be configured because the required %s env var is not present", controllers.CtrlNamespaceEnv)
+	}
+
+	// get the configuration secret
+	secret, created, err := getOrCreateConfigSecret(ctx, c, secretNamespace)
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			log.Info("kong configuration secret was created elsewhere retrying", "namespace", nsn.Namespace, "ingress", nsn.Name)
+			return ctrl.Result{Requeue: true}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	if created {
+		log.Info("kong configuration did not exist, was created successfully", "namespace", nsn.Namespace, "ingress", nsn.Name)
+		return ctrl.Result{Requeue: true}, nil
```

Comment:
Why is `Requeue: true` necessary here?

Reply:
Kubernetes controller best practices: Generally try to only make one write to the API per reconcilation. This function is re-entrant and so the next queue of it will reach the next logic gate.